stan/math/opencl/kernel_generator/binary_operation.hpp

```diff
@@ -0,0 +1,299 @@
+class subtraction_ : public binary_operation<subtraction_<T_a, T_b>, T_a, T_b> {
+ public:
+  /**
+   * Constructor.
+   * @param a first expression
+   * @param b second expression
+   */
+  subtraction_(T_a&& a, T_b&& b)  // NOLINT
+      : binary_operation<subtraction_<T_a, T_b>, T_a, T_b>(
+            std::forward<T_a>(a), std::forward<T_b>(b), "-") {}
+};
+
+/**
+ * Subtraction of two kernel generator expressions.
+ * @tparam T_a type of first expression
+ * @tparam T_b type of second expression
+ * @param a first expression
+ * @param b second expression
+ * @return Subtraction of given expressions
+ */
+template <typename T_a, typename T_b,
+          typename = require_all_valid_expressions_t<T_a, T_b>>
+inline subtraction_<as_operation_t<T_a>, as_operation_t<T_b>> operator-(
+    T_a&& a, T_b&& b) {  // NOLINT
+  return {as_operation(std::forward<T_a>(a)),
+          as_operation(std::forward<T_b>(b))};
+}
+
+/**
+ * Represents element-wise multiplication in kernel generator expressions.
+ * @tparam T_a type of first expression
+ * @tparam T_b type of second expression
+ */
+template <typename T_a, typename T_b,
+          typename = require_all_valid_expressions_t<T_a, T_b>>
+class elewise_multiplication_
+    : public binary_operation<elewise_multiplication_<T_a, T_b>, T_a, T_b> {
+ public:
+  /**
+   * Constructor.
+   * @param a first expression
+   * @param b second expression
+   */
+  elewise_multiplication_(T_a&& a, T_b&& b)  // NOLINT
+      : binary_operation<elewise_multiplication_<T_a, T_b>, T_a, T_b>(
+            std::forward<T_a>(a), std::forward<T_b>(b), "*") {}
+
+  /**
+   * View of a matrix that would be the result of evaluating this expression.
+   * @return view
+   */
+  inline matrix_cl_view view() const {
+    using base = binary_operation<elewise_multiplication_<T_a, T_b>, T_a, T_b>;
+    return both(std::get<0>(base::arguments_).view(),
+                std::get<1>(base::arguments_).view());
+  }
+};
+
+/**
+ * Element-wise multiplication of two kernel generator expressions.
+ * @tparam T_a type of first expression
+ * @tparam T_b type of second expression
+ * @param a first expression
+ * @param b second expression
+ * @return Element-wise multiplication of given expressions
+ */
+template <typename T_a, typename T_b>
+inline elewise_multiplication_<as_operation_t<T_a>, as_operation_t<T_b>>
+elewise_multiplication(T_a&& a, T_b&& b) {  // NOLINT
+  return {as_operation(std::forward<T_a>(a)),
+          as_operation(std::forward<T_b>(b))};
+}
+
+/**
+ * Multiplication of a scalar and a kernel generator expression.
+ * @tparam T_a type of scalar
+ * @tparam T_b type of expression
+ * @param a scalar
+ * @param b expression
+ * @return Multiplication of given arguments
+ */
+template <typename T_a, typename T_b, typename = require_arithmetic_t<T_a>,
+          typename = require_all_valid_expressions_t<T_b>>
+inline elewise_multiplication_<scalar_<T_a>, as_operation_t<T_b>> operator*(
```

Comment:
(apologies for the delay looking at this, just started back up at my jobby job)

Should this be `.*` to represent elementwise multiplication?

Reply:
This would be operator .* in matlab or similar languages. In c++ `operator.*` has [completely different meaning](https://stackoverflow.com/questions/746604/bind-pointer-to-member-operators-in-c/746626#746626) so I would not use it for element-wise multiplications.

Reply:
Ah word good call